In [1]:
#importing required libraries
import pandas as pd
import numpy as np


In [14]:
#Reading Data frame using Pd.read_excel
#using sheet_name parameter for reading a particular sheet in excel
#using header to remove unwanted empty rows and comment rows
#using header instead of skiprows because skip rows doesnt change the shape of data frame
Ac = pd.read_excel("Data Science Case study Vindiata.xls",sheet_name = "AC characteristics",header = 3)
Ac

Unnamed: 0 Aircraft Type  Range (Km)  Ave. Speed (km/h)  Number of Seats  \
0         NaN          A320        5000                800              150   
1         NaN          A330        8000                900              250   
2         NaN          B737        5000                800              150   
3         NaN          B747       10000                900              350   
4         NaN          Q400        1500                750               90   
5         NaN         ATR72        1000                650               75   

   Costs per flight hour  
0                   5000  
1                   7500  
2                   5100  
3                  12000  
4                   3500  
5                   2750

# Removing Unecesary Columns

In [19]:
Ac.drop(Ac.columns[0],axis=1,inplace=True)
Ac

Aircraft Type  Range (Km)  Ave. Speed (km/h)  Number of Seats  \
0          A320        5000                800              150   
1          A330        8000                900              250   
2          B737        5000                800              150   
3          B747       10000                900              350   
4          Q400        1500                750               90   
5         ATR72        1000                650               75   

   Costs per flight hour  
0                   5000  
1                   7500  
2                   5100  
3                  12000  
4                   3500  
5                   2750

In [20]:
Ac.shape

(6, 5)

# Viewing shape of dataframe
  
  Ac Dataframe consists of 6 rows and 5 columns  i.e 6 observations and 5 variables

# Reading Operations Sheet From Excel

In [32]:
#Reading Data frame using Pd.read_excel
#using sheet_name parameter for reading a particular sheet in excel
#using header to remove unwanted empty rows and comment rows
#using header instead of skiprows because skip rows doesnt change the shape of data frame
Op = pd.read_excel("Data Science Case study Vindiata.xls",sheet_name = "Operations",header = 3)

#Dropping Unnecesary columns
Op.drop(Op.columns[0], axis = 1, inplace = True)

Op

Aircraft Name Aircraft Type  2014-01-01 00:00:00  2014-02-01 00:00:00  \
0             A           A320                  250                  251   
1              B          A330                  300                  301   
2              C          B737                  245                  246   
3              D          B747                  310                  311   
4              E          Q400                  200                  201   
5              F         ATR72                  195                  196   
6              G          B747                  300                  301   
7              H          Q400                  195                  196   
8             I          ATR72                  200                  201   
9              J          A330                  310                  311   
10             K          B737                  250                  251   
11             L          A320                  245                  246   
12             M          A330                  300                  301   
13             N          B737                  245                  246   
14             O          A320                  250                  251   
15             P          A320                  245                  246   
16             Q          Q400                  195                  196   
17             R         ATR72                  200                  201   
18             S          A330                  310                  311   
19             T          A330                  300                  301   
20             U          B747                  310                  311   
21             V          Q400                  200                  201   
22             W         ATR72                  200                  201   
23             X          A320                  250                  251   
24             Y          A330                  300                  301   
25             Z          B737                  245                  246   

    2014-03-01 00:00:00  2014-04-01 00:00:00  2014-05-01 00:00:00  \
0                   248                  253                  243   
1                   298                  303                  293   
2                   243                  248                  238   
3                   308                  313                  303   
4                   198                  203                  193   
5                   193                  198                  188   
6                   298                  303                  293   
7                   193                  198                  188   
8                   198                  203                  193   
9                   308                  313                  303   
10                  248                  253                  243   
11                  243                  248                  238   
12                  298                  303                  293   
13                  243                  248                  238   
14                  248                  253                  243   
15                  243                  248                  238   
16                  193                  198                  188   
17                  198                  203                  193   
18                  308                  313                  303   
19                  298                  303                  293   
20                  308                  313                  303   
21                  198                  203                  193   
22                  198                  203                  193   
23                  248                  253                  243   
24                  298                  303                  293   
25                  243                  248                  238   

    2014-06-01 00:00:00  2014-07-01 00:00:00  2014-08-01 00:00:00  \
0                

In [33]:
#To view the sahpe of datafarame i.e (no.of rows , no.of columns)
Op.shape

(26, 14)

In [34]:
# Creating new column in Op 'Total Hours flown' i.e sum of hours of all months in a row
Op['Total hours flown'] = Op.iloc[:,2:].sum(axis=1)
Op


Aircraft Name Aircraft Type  2014-01-01 00:00:00  2014-02-01 00:00:00  \
0             A           A320                  250                  251   
1              B          A330                  300                  301   
2              C          B737                  245                  246   
3              D          B747                  310                  311   
4              E          Q400                  200                  201   
5              F         ATR72                  195                  196   
6              G          B747                  300                  301   
7              H          Q400                  195                  196   
8             I          ATR72                  200                  201   
9              J          A330                  310                  311   
10             K          B737                  250                  251   
11             L          A320                  245                  246   
12             M          A330                  300                  301   
13             N          B737                  245                  246   
14             O          A320                  250                  251   
15             P          A320                  245                  246   
16             Q          Q400                  195                  196   
17             R         ATR72                  200                  201   
18             S          A330                  310                  311   
19             T          A330                  300                  301   
20             U          B747                  310                  311   
21             V          Q400                  200                  201   
22             W         ATR72                  200                  201   
23             X          A320                  250                  251   
24             Y          A330                  300                  301   
25             Z          B737                  245                  246   

    2014-03-01 00:00:00  2014-04-01 00:00:00  2014-05-01 00:00:00  \
0                   248                  253                  243   
1                   298                  303                  293   
2                   243                  248                  238   
3                   308                  313                  303   
4                   198                  203                  193   
5                   193                  198                  188   
6                   298                  303                  293   
7                   193                  198                  188   
8                   198                  203                  193   
9                   308                  313                  303   
10                  248                  253                  243   
11                  243                  248                  238   
12                  298                  303                  293   
13                  243                  248                  238   
14                  248                  253                  243   
15                  243                  248                  238   
16                  193                  198                  188   
17                  198                  203                  193   
18                  308                  313                  303   
19                  298                  303                  293   
20                  308                  313                  303   
21                  198                  203                  193   
22                  198                  203                  193   
23                  248                  253                  243   
24                  298                  303                  293   
25                  243                  248                  238   

    2014-06-01 00:00:00  2014-07-01 00:00:00  2014-08-01 00:00:00  \
0                

In [35]:
#Creating a copy of Op i.e Operation for further use
Operations = Op.copy()

#Calculating total hours flown by each aircraft using groupby
Operations = Operations.groupby(['Aircraft Type']).sum()
Operations

2014-01-01 00:00:00  2014-02-01 00:00:00  2014-03-01 00:00:00  \
Aircraft Type                                                                  
A320                          1240                 1245                 1230   
A330                          1820                 1826                 1808   
ATR72                          795                  799                  787   
B737                           985                  989                  977   
B747                           920                  923                  914   
Q400                           790                  794                  782   

               2014-04-01 00:00:00  2014-05-01 00:00:00  2014-06-01 00:00:00  \
Aircraft Type                                                                  
A320                          1255                 1205                 1215   
A330                          1838                 1778                 1790   
ATR72                          807                  767                  775   
B737                           997                  957                  965   
B747                           929                  899                  905   
Q400                           802                  762                  770   

               2014-07-01 00:00:00  2014-08-01 00:00:00  2014-09-01 00:00:00  \
Aircraft Type                                                                  
A320                          1230                 1220                 1270   
A330                          1808                 1796                 1856   
ATR72                          787                  779                  819   
B737                           977                  969                 1009   
B747                           914                  908                  938   
Q400                           782                  774                  814   

               2014-10-01 00:00:00  2014-11-01 00:00:00  2014-12-01 00:00:00  \
Aircraft Type                                                                  
A320                          1270                 1245                 1250   
A330                          1856                 1826                 1832   
ATR72                          819                  799                  803   
B737                          1009                  989                  993   
B747                           938                  923                  926   
Q400                           814                  794                  798   

               Total hours flown  
Aircraft Type                     
A320                       14875  
A330                       21834  
ATR72                       9536  
B737                       11816  
B747                       11037  
Q400                        9476

In [36]:
#Joining two data frames using Merge to join both operations and Ac using common column=Aircraft Type
# we can use left or right or inner because we have same values for aircraft type
Ac_Op = Ac.merge(Operations , on = 'Aircraft Type' , how = 'left')
Ac_Op

Aircraft Type  Range (Km)  Ave. Speed (km/h)  Number of Seats  \
0          A320        5000                800              150   
1          A330        8000                900              250   
2          B737        5000                800              150   
3          B747       10000                900              350   
4          Q400        1500                750               90   
5         ATR72        1000                650               75   

   Costs per flight hour  2014-01-01 00:00:00  2014-02-01 00:00:00  \
0                   5000                 1240                 1245   
1                   7500                 1820                 1826   
2                   5100                  985                  989   
3                  12000                  920                  923   
4                   3500                  790                  794   
5                   2750                  795                  799   

   2014-03-01 00:00:00  2014-04-01 00:00:00  2014-05-01 00:00:00  \
0                 1230                 1255                 1205   
1                 1808                 1838                 1778   
2                  977                  997                  957   
3                  914                  929                  899   
4                  782                  802                  762   
5                  787                  807                  767   

   2014-06-01 00:00:00  2014-07-01 00:00:00  2014-08-01 00:00:00  \
0                 1215                 1230                 1220   
1                 1790                 1808                 1796   
2                  965                  977                  969   
3                  905                  914                  908   
4                  770                  782                  774   
5                  775                  787                  779   

   2014-09-01 00:00:00  2014-10-01 00:00:00  2014-11-01 00:00:00  \
0                 1270                 1270                 1245   
1                 1856                 1856                 1826   
2                 1009                 1009                  989   
3                  938                  938                  923   
4                  814                  814                  794   
5                  819                  819                  799   

   2014-12-01 00:00:00  Total hours flown  
0                 1250              14875  
1                 1832              21834  
2                  993              11816  
3                  926              11037  
4                  798               9476  
5                  803               9536

In [41]:
Ac_Op[['Aircraft Type','Range (Km)','Number of Seats','Costs per flight hour','Total hours flown']]

Aircraft Type  Range (Km)  Number of Seats  Costs per flight hour  \
0          A320        5000              150                   5000   
1          A330        8000              250                   7500   
2          B737        5000              150                   5100   
3          B747       10000              350                  12000   
4          Q400        1500               90                   3500   
5         ATR72        1000               75                   2750   

   Total hours flown  
0              14875  
1              21834  
2              11816  
3              11037  
4               9476  
5               9536

In [42]:
# Findinng the total cost by aircraft type for the year
# Total cost = Total hours flown * Costs per flight hour

Ac_Op['Total Cost'] = Ac_Op['Total hours flown'] * Ac_Op['Costs per flight hour']
Ac_Op[['Aircraft Type','Total hours flown','Total Cost']]

Aircraft Type  Total hours flown  Total Cost
0          A320              14875    74375000
1          A330              21834   163755000
2          B737              11816    60261600
3          B747              11037   132444000
4          Q400               9476    33166000
5         ATR72               9536    26224000

In [45]:
#Viewing required columns and sorted by total cost
Ac_Op[['Aircraft Type','Range (Km)','Number of Seats','Costs per flight hour','Total hours flown','Total Cost']].sort_values(['Total Cost'],ascending=False)

Aircraft Type  Range (Km)  Number of Seats  Costs per flight hour  \
1          A330        8000              250                   7500   
3          B747       10000              350                  12000   
0          A320        5000              150                   5000   
2          B737        5000              150                   5100   
4          Q400        1500               90                   3500   
5         ATR72        1000               75                   2750   

   Total hours flown  Total Cost  
1              21834   163755000  
3              11037   132444000  
0              14875    74375000  
2              11816    60261600  
4               9476    33166000  
5               9536    26224000

# We can see the Total Cost of every Aircraft type for the year

- Here A330 has flown more hours and it has the higest cost
- B747 has more capacity in seating and Range it stands second in cost 
- ATR72 stands last in Cost but it has very special features like seating capacity is 75 and Range is 1000 Km we can use on those conditions satisfy it is most useful for short trips with less people

# Which aircraft type  has the lowest cost per seat per km flown?

- Cost per seat Per Km = Total Cost / (no.of seats * total hours flown * Average Speed )

In [48]:
# Creating new column Cost per seat per Km using above formula
Ac_Op['Cost per seat per Km'] = Ac_Op['Total Cost'] / (Ac_Op['Number of Seats']*Ac_Op['Ave. Speed (km/h)']*Ac_Op['Total hours flown'])

Ac_Op[['Aircraft Type','Cost per seat per Km']].sort_values(['Cost per seat per Km'])

Aircraft Type  Cost per seat per Km
1          A330              0.033333
3          B747              0.038095
0          A320              0.041667
2          B737              0.042500
4          Q400              0.051852
5         ATR72              0.056410

# Here we can See that A330 has the lowest cost per seat per Km
- it is the most econmical Aircraft Among all 


In [51]:
#Reading Data frame using Pd.read_excel
#using sheet_name parameter for reading a particular sheet in excel
#using header to remove unwanted empty rows and comment rows
#using header instead of skiprows because skip rows doesnt change the shape of data frame
Cp = pd.read_excel("Data Science Case study Vindiata.xls",sheet_name = "City pairs",header = 3)

#Dropping Unnecesary columns
Cp.drop(Cp.columns[0], axis = 1, inplace = True)
Cp

Origin City Desitnation City  Pass. Demand \n(per day)  Distance (km)
0          AA               BB                       420           3000
1          BB               CC                       450           6500
2          CC               AA                       300            400
3          AA               DD                       300           1000

In [ ]:
#Creating a Copy of Ac_Op as df for futher evaluation
df = Ac_Op.copy()

#Dropping columns that doesnt require for Evaluation
df.drop(df.iloc[:, 5:19], axis = 1, inplace = True)
df

In [56]:
#Dropping columns that doesnt require for Evaluation
df.drop(df.columns[[2,4]] , axis = 1 , inplace = True)
df

Aircraft Type  Range (Km)  Number of Seats  Cost per seat per Km
0          A320        5000              150              0.041667
1          A330        8000              250              0.033333
2          B737        5000              150              0.042500
3          B747       10000              350              0.038095
4          Q400        1500               90              0.051852
5         ATR72        1000               75              0.056410

# Finding the Best Aircrafts for Different city Pairs
- First we check whether aircraft range is greater than the Distance between Citys
- Then we calculate Cost for trip



In [60]:
# total cost of trip = cost per seat per Km * no.of Seats * no.of trips * Trip distance
# no.of trips = pass demand per day/no.of seats
# Creating new columns in df using for loop using itertuple
# if condition doesnt satisfy it will give "NaN"

for a,b,c,d in Cp.itertuples(index=False):
    df[str(a)+'-'+str(b)+' cost'] = np.where(df['Range (Km)'] >= d, (df['Cost per seat per Km'] * df['Number of Seats'] * d * np.ceil(c / df['Number of Seats'])).astype(int), "NaN")

# Display the dataframe with added travel cost
df

Aircraft Type  Range (Km)  Number of Seats  Cost per seat per Km AA-BB cost  \
0          A320        5000              150              0.041667      56250   
1          A330        8000              250              0.033333      50000   
2          B737        5000              150              0.042500      57375   
3          B747       10000              350              0.038095      80000   
4          Q400        1500               90              0.051852        NaN   
5         ATR72        1000               75              0.056410        NaN   

  BB-CC cost CC-AA cost AA-DD cost  
0        NaN       5000      12500  
1     108333       6666      16666  
2        NaN       5100      12750  
3     173333       5333      13333  
4        NaN       7466      18666  
5        NaN       6769      16923

# Most Suitable Economical Aircraft type for every City pair with number of trips

- AA-BB A330 2 trips
- BB-CC A330 2 trips
- CC-AA A320 2 trips
- AA-DD A320 2 trips


In [61]:
Cp

Origin City Desitnation City  Pass. Demand \n(per day)  Distance (km)
0          AA               BB                       420           3000
1          BB               CC                       450           6500
2          CC               AA                       300            400
3          AA               DD                       300           1000

# From the above 2 tables df , Cp
- **From AA-BB distance = 3000 , person demand per day = 420**
  - for aircraft A320 Range=5000 , No.of seats = 150 
    - It takes 3 trips i.e 420/150 =~ 3
  - for air craft A330 Range=8000 , no.of seats = 250
    - It takes 2 trips i.e 420/250 =~ 2
  - for aircraft B737 Range=5000 , No.of seats = 150 
    - It takes 3 trips i.e 420/150 =~ 3
  - for air craft B747 Range=10000 , no.of seats = 350
    - It takes 2 trips i.e 420/350 =~ 2  
  - Other aircrafts range not sufficient for the distance 

# Among the Capable air Crafts **A330** is more Economical we can see in above table i.e **Rs.50000**
  




- **From BB-CC distance = 6500 , person demand per day = 450**
  - for air craft A330 Range=8000 , no.of seats = 250
    - It takes 2 trips i.e 450/250 =~ 2
  - for air craft B747 Range=10000 , no.of seats = 350
    - It takes 2 trips i.e 450/350 =~ 2  
  - Other aircrafts range not sufficient for the distance 


# Among the Capable air Crafts **A330** is more Economical we can see in above table i.e **Rs.108333**  
  





- **From CC-AA distance = 400 , person demand per day = 300**
  - for aircraft A320 Range=5000 , No.of seats = 150 
    - It takes 2 trips i.e 300/150 =~ 2
  - for air craft A330 Range=8000 , no.of seats = 250
    - It takes 2 trips i.e 300/250 =~ 2
  - for aircraft B737 Range=5000 , No.of seats = 150 
    - It takes 2 trips i.e 300/150 =~ 2
  - for air craft B747 Range=10000 , no.of seats = 350
    - It takes 1 trips i.e 300/350 =~ 1  
  - for aircraft Q400 Range=1500 , No.of seats = 90 
    - It takes 4 trips i.e 300/90 =~ 4
  - for air craft ATR72 Range=1000 , no.of seats = 75
    - It takes 4 trips i.e 300/75 =~ 4 

# Among the Capable air Crafts **A320** is more Economical we can see in above table i.e **Rs.5000**



# We Can also Consider Aircraft **B747** is second most Economical i.e **RS.5333** but can be done in A single trip time and Energy both can be saved
  




- **From AA-DD distance = 1000 , person demand per day = 300**
  - for aircraft A320 Range=5000 , No.of seats = 150 
    - It takes 2 trips i.e 300/150 =~ 2
  - for air craft A330 Range=8000 , no.of seats = 250
    - It takes 2 trips i.e 300/250 =~ 2
  - for aircraft B737 Range=5000 , No.of seats = 150 
    - It takes 2 trips i.e 300/150 =~ 2
  - for air craft B747 Range=10000 , no.of seats = 350
    - It takes 1 trips i.e 300/350 =~ 1  
  - for aircraft Q400 Range=1500 , No.of seats = 90 
    - It takes 4 trips i.e 300/90 =~ 4
  - for air craft ATR72 Range=1000 , no.of seats = 75
    - It takes 4 trips i.e 300/75 =~ 4 


# Among the Capable air Crafts **A320** is more Economical we can see in above table i.e **Rs.12500**


# We Can also Consider Aircraft **B747** is second most Economical i.e **RS.13333** but can be done in A single trip time and Energy both can be saved  
  
  
  
    